### ISSUE TITLE
Can't bind controller trigger axis events (L2, R2) as if buttons (4.0 beta)

### ISSUE THREAD

**Issue by the-eclectic-dyslexic** [godotengine/godot#72636]
### Godot version

4.0 beta (<=17)

### System information

Arch Linux

### Issue description

In godot 3.5, it was possible to bind a trigger (L2 or R2) event to a `InputEventJoypadButton`. (Through what was `button_index` 6 and 7) this allowed for watching for a trigger press via `InputEvent.is_action_pressed("action")` and `InputEvent.is_action_released("action")`

However, in godot 4 beta (<=17), this functionality has been removed, and you can only bind the triggers as `InputEventJoypadMotion` (axis 4 and 5). You may still monitor events through  `InputEvent.is_action_pressed("action")` and `InputEvent.is_action_released("action")`... However, multiple events will be sent, instead of just one like is done with a button or a key without echo events. This seems to be expected behaviour... (other than that i notice `InputEvent.is_action_released("action")` is  true while slowly SQUEEZING the trigger, not when releasing it.. the only thing I can think there is that whether an event is being pressed or released is being defined by the axis strength < 0.5 as opposed to the axis strength being changed relative to what it was last it was checked.)

A workaround to this behaviour is to use `Input.is_action_just_pressed("action")` and `Input.is_action_just_pressed("action")` instead. However, this is not a complete solution. As this cannot be run inside `_input(event: InputEvent)` given that it seems the Input singleton is only updated once per process frame I think? Meaning that if you want to handle trigger inputs, and want to allow rebinding of controller inputs... you cannot use the `_input(event: InputEvent)` function AT ALL, and must watch all of your inputs through the `Input` Singleton inside of `_process()`

As I see it, there are two solutions to this problem:

1. bring back the option to bind (L2 and R2) as `InputEventJoypadButton` using the same rules they followed in 3.5

2. change the behaviour of `InputEventJoypadMotion.is_action_pressed("action")` to check for having just left the defined deadzone and `InputEventJoypadMotion.is_action_released("action")` to check for having just entered the defined deadzone. Though, I am not sure this would be a desirable change for anyone other than myself, as it would limit the use of InputEvent monitoring through `_input(event: InputEvent)` in a different way.

### Steps to reproduce

I have attached two projects, that try to implement the exact same thing. One is in godot 3.5, while the other uses the 4.0 beta.

You can see the behaviour as it was in 3.5, by using L2 (button 6) which is bound as a `InputEventJoypadButton`
In the 4.0 beta version L2 is not bound, as I can't find a way to bind it as an `InputEventJoypadButton`... and as I said above, I suspect that functionality no longer exists. The script tries looking at all button indexes after 15 to see if any of them happen to bind to L2, and none seem to.

In both versions R2 (axis 7 in 3.5, axis 5 in 4.0 beta ) is bound as an `InputEventJoypadAxis`. This allows looking at an example of the workaround I mention above.

In both you can change from reading input through `_input(event: InputEvent)` to `_process()` and back by pressing button 0 (x on a sony controller)

A parity count is kept based on the number of times a button presses or releases for testing purposes, and each project will print the count along with some other information every time the trigger is pressed or released.

### Minimal reproduction project

[triggers_3_5.zip](https://github.com/godotengine/godot/files/10574279/triggers_3_5.zip)
[triggers_4_beta_17.zip](https://github.com/godotengine/godot/files/10574280/triggers_4_beta_17.zip)

**Comment by feiyunw** [comment]
I see no reason to abandon L2 and R2 button bindings. I use them in my game!

FYI: [DUALSHOCK&trade;4 wireless controller](https://manuals.playstation.net/document/en/ps4/basic/pn_controller.html)

**Comment by Nathan-R-Og** [comment]
I came across this problem while porting my input profiler to Godot 4. I'm sure a workaround can be done by having a custom parser that checks if a trigger is fully pressed or not, and sends a InputEventJoypadMotion or InputEventJoypadButton respectively through different actions. However, I definitely should not have to go out of my way to manually parse inputs just because one rebindable input is now removed. Either this is an unexpectedly weird known change, or this needs to get fixed.
For the time being, I'm sure it'll be fine for the majority of projects to go without a trigger click, but to have the option is greatly appreciated.
I'm using an XBOX XInput controller, and as of Godot 3.4.1 stable (the original project the profiler came from), the trigger clicks worked as intended.

**Comment by the-eclectic-dyslexic** [comment]
@Nathan-R-Og Ya, you could write your own input system that you would use ontop of Godot's action system just for 2 buttons... but I think that is a little bit much to be expected as I wouldn't really consider that a solution so much as a reliable way to create bugs. Even if your wrapper works perfectly, you need to wrap inputs in every single location that you want to handle inputs. It might be more reasonable if you could place your wrapper directly in the way when the engine needs to call the _input or _unhandled_input methods. Which I am sure could be done with an extension of the engine... but this is kind of ridiculous.

I will readily admit that this issue in particular has been a real thorn in my side. Writing the logic for controller rebinding and only getting to use the Input singleton everywhere I need inputs has not been a good time. For my purposes, the _input and _unhandled_input functions may as well not exist... despite them just being the better option from a code organization standpoint imo.

**Comment by hunterloftis** [comment]
@KoBeWi is this really a discussion rather than regression?

I just ran into this surprising bug myself, and one thing that makes it extra troublesome is that for some cases it will mostly seem to work. It _does_ trigger `is_action_pressed` in `_unhandled_input`, but it triggers it multiple times and in surprising orders. The only reason I went hunting for a bug in github is because my binding to a "regular" button was working as expected, and only activating the same behaviors with triggers was breaking in weird ways.

**Comment by hunterloftis** [comment]
If anyone else is looking for a workaround, I've added this as an autoload to re-create the Godot 3.5 behavior in my game. Set up empty input maps (in this case, "use" and "dash"). Then set up input maps that map to the triggers ("use_trigger", "dash_trigger"). Then this will map the trigger actions to the correct input events:

```gdscript
extends Node

func _process(delta: float) -> void:
	if _fix_trigger_action("use_trigger", "use"): return
	if _fix_trigger_action("dash_trigger", "dash"): return
        # ...plus any other inputs you need to fix
	
func _fix_trigger_action(trigger_action: String, action: String) -> bool:
	if Input.is_action_just_pressed(trigger_action):
		_simulate_action(action, true)
		return true
	if Input.is_action_just_released(trigger_action):
		_simulate_action(action, false)
		return true
	return false
		
func _simulate_action(action: String, pressed: bool) -> void:
	var event = InputEventAction.new()
	event.action = action
	event.pressed = pressed
	Input.parse_input_event(event)
```

**Comment by KoBeWi** [comment]
As I said, since triggers are now axes (which is more correct, because they aren't boolean input), they work as such and need to be handled differently. Left/right stick actions will yield same/similar results in `_unhandled_input()`, because axis input is like that. The solution is not using third-party addons but adjusting the code. This works perfectly fine:
```GDScript
func _process(delta):
    if Input.is_action_just_pressed("L2"):
        print("trigger pressed")
```
How do you propose this should be fixed? Triggers can't send both Button and Axis events, because it would be inconsistent.

**Comment by hunterloftis** [comment]
From the [Godot docs](https://docs.godotengine.org/en/stable/tutorials/inputs/controllers_gamepads_joysticks.html#supporting-universal-input):

> Thanks to Godot's input action system, Godot makes it possible to support both keyboard and controller input without having to write separate code paths.

More from the [Godot docs](https://docs.godotengine.org/en/stable/tutorials/inputs/input_examples.html#capturing-actions):

> Once you've defined your actions, you can process them in your scripts using is_action_pressed() and is_action_released() by passing the name of the action you're looking for
> func _input(event):
>    if event.is_action_pressed("my_action"):
>        print("my_action occurred!")

Dropping support for triggers as discrete inputs for actions breaks that promise. It's no longer possible to say: "space is shoot and trigger is shoot, so map both of those to the 'shoot' action and then handle that action in `_input` as described in the Godot docs." That would work in Godot 3, but not in 4.

While I understand the argument of "it's more pure" and "it's *technically* correct," one of the selling points of Godot so far has been that you don't need extra code to handle the most common types of input, and you can even support the most-common cases of keyboard + controller support universally. This departure from that behavior breaks that promise, and breaks that universality.

> How do you propose this should be fixed? Triggers can't send both Button and Axis events, because it would be inconsistent.

The original poster of this issue proposed two valid paths.

But that's just my opinion as a user, and of course I can't expect you to agree with me. However, I would ask for clear communication: if the project is set on *not* supporting trigger buttons as universal inputs, would you please make that clear and close this issue? Then instead of adding one-off workarounds to every Godot 4 project, users like me can build the library that will be needed to bridge the gap - something similar to [this](https://github.com/godotengine/godot/issues/72636#issuecomment-1676488169) - and move on.

**Comment by KoBeWi** [comment]
The fact that triggers are axes is not going to change (aside from the reasoning above, it would also break compatibility with the previous 4.x versions). We need to think of some solution to resolve this issue. The suggestions here will be considered, but the actual solution might differ.

**Comment by the-eclectic-dyslexic** [comment]
@KoBeWi I think I mostly agree with hunter on this one. The way the triggers are treated for the purposes of `InputEvent.IsActionPressed("action")` don't make any sense to me, so much so that I thought I just didn't understand some method behind the madness, and didn't want to suggest to strongly about changing the behaviour in case someone understood something I didn't and I broke a bunch of people's projects who were relying on the seemingly arcane behaviour. In fact they make so little sense I see now that I failed to describe the utter confusion I had when I came across this behavour. That was the purpose of the parity counter in the provided projects. Holding the trigger the right way can easily get you dozens of press actions and zero release actions... or vice versa. 

As per above, I am actually much more in favour of `InputEvent.IsActionPressed("action")` being only true when the deadzone has just been left. Not doing this could be considered okay by me only if some kind of reasoning can be given behind the behaviour as is... additionally that behaviour should be well documented so that people actually know what it does and why. If that doesn't happen, people are going to do what I did when I posted this issue (trying to use it assuming what I suggested in option 2 of the original post was the behaviour) and then be utterly perplexed as to what is going on when they rebind something to it and find it utterly broken by that behaviour... when the action they had bound by default worked fine despite the strange behaviour. The alternate option of "no you cannot press an axis" also seems fine to me, if no reason for the behaviour can be given, and a different method is used to tackle this issue as you suggested by "the actual solution may differ".

Also...

@hunterloftis That code snippet is an unreasonably KISS solution, and I am salty I didn't think of it and  worked around it in a hugely verbose way instead. I need to up my godot-fu! I will note though... If I am reading it correctly, I think that will only fix one button per frame. If you press both triggers during the same frame, I don't think your solution will work as intended. It will only fix your "use_trigger" in that case. Which may be fine for your game, but it is something to consider if you run into a bug with your "dash_trigger" not registering in the future.

Also, thank you for making a stink and bringing attention to this issue!

**Comment by hunterloftis** [comment]
@the-eclectic-dyslexic good catch, those early returns that I thought were optimizations were subtle bugs! Fixed, thank you.

I had to check whether or not pressed and released can both fire during the same frame, or if Godot queues them or something, and learned that until recently they could not:

https://github.com/godotengine/godot/pull/77055

**Comment by Nathan-R-Og** [comment]
I guess I might as well throw my hat into the ring again; For my profiler, I use _exclusively_ InputEvent objects, so that they may be stored to actions and any amount of actions.
I'd rather not make an entirely separate action for trigger clicks, as that more or less defeats the point of using actions as game-oriented control handlers. Every profile would need to have an added action to accommodate, and I don't think it's worth it.
As for triggers being axes instead of button presses, that makes total sense! The axis give plenty more range and are (probably) way more versatile. I just wonder, what's stopping us from having both?
It was mentioned that the left and right analog sticks also have this changed, but this doesn't seem to be the case.
In my profiler, I was able to detect if it was a valid analog-direction-flick with a threshold, and the stick click still works just as fine. Most of the time (unless you need INCREDIBLY precise data), you can just make a direction vector out of Input.get_vector(action_left,action_right,action_up,action_down) anyways.
Those are just my two cents anyhow.

**Comment by idlist** [comment]
> Triggers can't send both Button and Axis events, because it would be inconsistent.

Imagine you have a function to let the player configure what button they would like to use for certain actions (which is not uncommon nowadays, especially with games that does not need to treat L2 and R2 as axis), and if you want to support having L2 and R2 be able to set in this scenario, you **have to** use `Input.is_action_just_pressed()` everywhere in contrast with the tutorial or the documents, where `event.is_action_pressed()` inside `_unhandled_input(event)` is used elsewhere. **This** in my opinion is the true inconsistency imposed by this issue.

**Comment by Lunarthewolf** [comment]
@Nathan-R-Og I have seen this issue with the analog stick as well. I was trying to implement a "double tap to run" feature and my character was sometimes running with a single tap. My workaround was to wait until the deadzone reaches an even lower point than the threshold before accepting another press:

```gdscript
var analog_actions_pressed := {
	"move_left_analog" = false,
	"move_right_analog" = false,
	"move_up_analog" = false,
	"move_down_analog" = false
}

func _unhandled_input(event: InputEvent):
	for analog_action in analog_actions_pressed.keys():
		if event.is_action_pressed(analog_action) && !analog_actions_pressed[analog_action]:
			analog_actions_pressed[analog_action] = true
			# handler code goes here
		elif Input.get_action_raw_strength(analog_action) < 0.1:
			analog_actions_pressed[analog_action] = false
```

Could something like this be implemented as a back end solution?

**Comment by SphyxDev** [comment]
As this issue is still open as of 4.4 beta 1, I wanted to share my perspective as a user, being unfamiliar with the engine source code. The vast majority of projects would presumably treat the trigger buttons no differently than bumper buttons: as digital buttons. That is, whether or not the controller has analog triggers, they would behave like digital triggers. The fact that we currently can't just use `InputEvent.is_action_pressed("action")` under `_input()` or `_unhandled_input()` for this behavior if the controller happens to have analog triggers is very inconvenient. Especially when considering allowing the end user to remap an action to any button, including the triggers, which is a very common scenario.

As for the projects that do use analog triggers, we can already get the input strength using `InputEvent.get_action_strength("action")`.  From there, it would be on the developer to use this value with regards to what the project needs. (e.g. save the [0, 1] value to some variable x and add (x * SOME_CONSTANT) to a car's speed.)

In this case, I believe simply allowing reading the input like a digital button with `InputEvent.is_action_pressed("action")` would be the expected result for most users. It sounds the most practical to me anyway, instead of being consistent for the sake of consistency.

**Comment by sergioabreu-g** [comment]
Why is this being discussed at all? Treating axes as buttons is done all the time in games.

Using `Input.is_action_just_pressed("action")` instead doesn't make any sense. How do you handle button remapping then? Are you going to tell the user they can't bind a trigger to an action because it's an axis? Or are you going to refactor all of your inputs to use `is_action_just_pressed` just to support axes inputs? Then, why is there an `input` function at all? We should check inputs in the `process` function every time to avoid any potential issues with axes.

**Comment by Zain-A-Abbas** [comment]
Indeed, I found this thread while trying to resolve an issue with my game's input remapping, and the current way Godot treats triggers is very misaligned from how the vast majority of games treat them: as buttons. If you want to allow players to remap your controls, you _have_ to engineer a completely different solution for how the game treats InputEvents that happen to be bound to an axis.

If anyone else has stumbled across this thread, here is the workaround I am using:

```gdscript
extends Node

const THRESHOLD: float = 0.5

var passed_thresholds: Dictionary[JoyAxis, bool] = {
	JoyAxis.JOY_AXIS_TRIGGER_LEFT: false,
	JoyAxis.JOY_AXIS_TRIGGER_RIGHT: false,
	JoyAxis.JOY_AXIS_LEFT_X: false,
	JoyAxis.JOY_AXIS_LEFT_Y: false,
	JoyAxis.JOY_AXIS_RIGHT_X: false,
	JoyAxis.JOY_AXIS_RIGHT_Y: false,
}

var deviceId: int = 0

func _process(delta: float) -> void:
	for axis in passed_thresholds.keys():
		var axis_value: float = Input.get_joy_axis(deviceId, axis)
		if axis_value <= THRESHOLD && passed_thresholds[axis]:
			passed_thresholds[axis] = false

func set_device_id(event: InputEvent):
	if event is InputEventJoypadButton || event is InputEventJoypadMotion:
		deviceId = event.device

func is_pressed(event: InputEvent):
	set_device_id(event)
	var pressed: bool = event.is_pressed()
	return final_value(event, pressed)

func is_action_pressed(event: InputEvent, action: StringName, allow_echo: bool = false, exact_match: bool = false):
	set_device_id(event)
	var pressed: bool = event.is_action_pressed(action, allow_echo, exact_match)
	return final_value(event, pressed)

func final_value(event: InputEvent, actual_value: bool) -> bool:
	if !actual_value:
		return false
	if event is InputEventJoypadMotion:
		if event.axis_value > THRESHOLD && passed_thresholds[event.axis] == false:
			passed_thresholds[event.axis] = true
			return actual_value
		return false
	
	return actual_value
	
```

This autoload is called in place of _all_ instances of `event.is_action_pressed()`, with the idea that it does not return true on joypad axis inputs until they pass a certain threshold _and_ were not already in that threshold the previous frame.

**Comment by aleksa2808** [comment]
- I initially did my input via `Input.is_action_just_pressed()` which worked OK.
- Then my controller stopped working for a bit.
- In the meantime I found out I should use the `_input()` functions instead as they handle things such as GUIs better, so I reimplemented my input code.
- Then I fixed my controller and found out that it doesn't work anymore for my game as both the joystick inputs (which cover `ui_up`/`ui_down`) as well as the R2/L2 triggers get treated as axes and keep spamming the `_input` functions with events that get caught by `is_action_pressed` calls.

I'm at a loss here, how SHOULD we handle input then? Implementing a wrapper class that all input goes through just so we can treat these inputs as binary, like some here suggest, feels like an anti-pattern.
